### PR TITLE
Boards: add nz32-sc151

### PR DESCRIPTION
**Pull request by fjmolinas**
Hi,
nz32-sc151 board has been ported to RIOT, it has been tested on a couple of standard examples such as: hello-world, default. The different peripherals have passed there respective test's. 

Also, dac and adc support has been added to the stm32l1 cpu used in the nz32-sc151. It has been tested using the periph_dac and periph_adc tests only on this board. This content can be found in PR #6383.

Flashing is done via dfu, board must manually be put into bootloader via user access buttons. The usb port doesn't open a uart communication therefore a usb/ttl converter must be directly connected to the board to open a term.

### CODE REVIEW

**Comment by kYc0o** [comment]
Hi @fjmolinas ! I see that your PR didn't have a review yet... We will try to get the hardware to test and then make a proper review. For now some "minor" remarks:

1. PR names usually, but not mandatory, follow a scheme of "boards: add ... " or "pkg: add/modify/fix ..." so maybe we could change the name of this PR.

2. Commits also follow a name scheme: "boards: add nz32-sc151 board" for a commit which is adding/changing files on such path. The same for all files you're adding/modifying.

3. You can add commits and preferably squash them until a review starts, so we can work on a clean branch since the beginning. Once the reviewer asks for a change you don't squash it until it's validated by a maintainer.

So for now, what I'd like to have is some commits like:

boards: add nz32-sc151
cpu/stm32l: add adc and dac

But as far as I can see, this second commit should be in a separated PR, since it adds a feature to an existing cpu.

Thanks for your contribution and I'll try to make a proper review soon! (at a first sight everything looks good so I think it can be quick)

**Comment by fjmolinas** [comment]
Hi @kYc0o, sorry for not getting back to you sooner, haven't had time to get back to this PR. Would it be preferable to do a new PR on a cleaner branch so that commits follow the schema you are asking for? It would be hard to squash it all clean since I merged with master's updates in between witch forced some style changes and others. Cheers!

**Comment by kYc0o** [comment]
Uhm... I think this PR is still healthy but as you prefer. Don't forget to split it in two.

**Comment by fjmolinas** [comment]
Hi @kYc0o, I spplited them into two different PR, the first one being #6383. Of course they are dependent since this PR uses features from 6383. Cheers!

**Comment by kYc0o** [comment]
Just got the board, I can test if you make another PR with only this board added. You can also modify this one with only the board.

File: boards/nz32-sc151/include/board.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * @brief       Board specific files for the nz32-sc151 board.
+ * @{
+ *
+ * @file
+ * @brief       Board specific definitions for the nz32-sc151 board.
+ *
+ * @author      Francisco Molina <francisco.molina@inria.cl>
+ */
+
+#ifndef BOARD_H_
```
Comment (aabadie) [inline boards/nz32-sc151/include/board.h]: trailing underscore, please remove. You should have:
```c
#ifndef BOARD_H
```
instead

File: boards/nz32-sc151/include/board.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * @brief   USER LED pin definitions and handlers
+ * @{
+ */
+#define LED0_PIN            GPIO_PIN(PORT_B, 2)
+
+#define LED0_PORT           (GPIOB)
+#define LED0_MASK           (1 << 2)
+
+#define LED0_ON             (LED0_PORT->BSRRL = LED0_MASK)
```
Comment (aabadie) [inline boards/nz32-sc151/include/board.h]: The BSRRL and BSRRH register have been remove from the CMSIS, now there's only one 32bit register BSRR.
```c
#define LED0_ON             (GPIOB->BSRR = LED0_MASK)
#define LED0_OFF            (GPIOB->BSRR = (LED0_MASK << 16))
#define LED0_TOGGLE         (GPIOB->ODR  ^= LED0_MASK)
```

File: boards/nz32-sc151/include/board.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+extern "C" {
+#endif
+
+/**
+ * @brief   USER LED pin definitions and handlers
+ * @{
+ */
+#define LED0_PIN            GPIO_PIN(PORT_B, 2)
+
+#define LED0_PORT           (GPIOB)
```
Comment (aabadie) [inline boards/nz32-sc151/include/board.h]: you can remove this define

File: boards/nz32-sc151/include/board.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * @brief Initialize board specific hardware, including clock, LEDs and std-IO
+ */
+void board_init(void);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* BOARD_H_ */
```
Comment (aabadie) [inline boards/nz32-sc151/include/board.h]: trailing underscore + missing newline at end of file

File: boards/nz32-sc151/include/periph_conf.h [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+ * @ingroup     nz32-sc151
+ * @{
+ *
+ * @file
+ * @brief       Peripheral MCU configuration for the limifrog-v1 board
+ *
+ * @author      Francisco Molina <francisco.molina@inria.cl>
+ */
+
+#ifndef PERIPH_CONF_H_
```
Comment (aabadie) [inline boards/nz32-sc151/include/periph_conf.h]: underscore

File: boards/nz32-sc151/include/periph_conf.h [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    {
+        .dev      = USART3,
+        .rcc_mask = RCC_APB1ENR_USART3EN,
+        .rx_pin   = GPIO_PIN(PORT_B, 11),
+        .tx_pin   = GPIO_PIN(PORT_B, 10),
+        .rx_af    = GPIO_AF7,
+        .tx_af    = GPIO_AF7,
+        .bus      = APB1,
+        .irqn     = USART3_IRQn
+    }/*,
```
Comment (aabadie) [inline boards/nz32-sc151/include/periph_conf.h]: why commenting ? are the extra UART not working ?

File: boards/nz32-sc151/include/periph_conf.h [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+#define UART_1_ISR          (isr_usart2)
+#define UART_2_ISR          (isr_usart1)
+
+#define UART_NUMOF          (sizeof(uart_config) / sizeof(uart_config[0]))
+
+/**
+ * @brief SPI configuration
+ * @{
+ */
+#define SPI_NUMOF           (SPI_0_EN + SPI_1_EN)
```
Comment (aabadie) [inline boards/nz32-sc151/include/periph_conf.h]: The SPI definition needs to be updated

File: boards/nz32-sc151/include/periph_conf.h [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+#define I2C_1_ERR_ISR       isr_i2c2_er
+
+static const i2c_conf_t i2c_config[] = {
+    /* device, port, scl-, sda-pin-number, I2C-AF, ER-IRQn, EV-IRQn */
+    {I2C1, GPIO_PIN(PORT_B,  8), GPIO_PIN(PORT_B,  9),
+     GPIO_OD_PU, GPIO_AF4, I2C1_ER_IRQn, I2C1_EV_IRQn},
+    {I2C2, GPIO_PIN(PORT_B, 10), GPIO_PIN(PORT_B, 11),
+     GPIO_OD_PU, GPIO_AF4, I2C2_ER_IRQn, I2C2_EV_IRQn},
+};
+
```
Comment (aabadie) [inline boards/nz32-sc151/include/periph_conf.h]: newline that can be removed

File: boards/nz32-sc151/include/periph_conf.h [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+ * @{
+ */
+#define RTC_NUMOF           (1U)
+/** @} */
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* PERIPH_CONF_H_ */
```
Comment (aabadie) [inline boards/nz32-sc151/include/periph_conf.h]: underscore + newline at end of file

File: cpu/stm32l1/include/periph_cpu.h [hunk 10]
```diff
@@ -94,4 +139,3 @@ typedef struct {
 #endif
 
 #endif /* PERIPH_CPU_H */
-/** @} */
```
Comment (aabadie) [inline cpu/stm32l1/include/periph_cpu.h]: please restore the doxygen comment

File: cpu/stm32l1/include/stm32l1xx.h [hunk 11]
```diff
@@ -6671,4 +6670,4 @@ typedef struct
   * @}
   */
 
-/************************ (C) COPYRIGHT STMicroelectronics *****END OF FILE****/
+/************************ (C) COPYRIGHT STMicroelectronics *****END OF FILE****/
```
Comment (aabadie) [inline cpu/stm32l1/include/stm32l1xx.h]: you removed the newline at end of file

File: cpu/stm32l1/periph/timer.c [hunk 12]
```diff
@@ -0,0 +1,163 @@
+/*
```
Comment (aabadie) [inline cpu/stm32l1/periph/timer.c]: timer implementation has moved in a common place. Can you revert this change ?

File: cpu/stm32l1/periph/dac.c [hunk 13]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (aabadie) [inline cpu/stm32l1/periph/dac.c]: the dac periph implementation has moved to a common place. Can you revert this change ?

File: cpu/stm32l1/periph/adc.c [hunk 14]
```diff
@@ -0,0 +1,133 @@
+/*
```
Comment (aabadie) [inline cpu/stm32l1/periph/adc.c]: This file should be part of a separate PR as discussed here before.

File: examples/hello-world/main.c [hunk 15]
```diff
@@ -23,6 +23,7 @@
 
 int main(void)
 {
+
```
Comment (aabadie) [inline examples/hello-world/main.c]: please revert

**Comment by fjmolinas** [comment]
@aabadie Hi, sorry for the long absence I have been on vacation and then having to deal with college finals. I was planing on getting back to this PR, but right now I don't have access to the hardware anymore. There for I couldn't really make sure any changes were properly working. I was planing on doing all the modification I could this weekend, but I won't be able to properly test it.

Answering your question on the DFU connection, I never used the ST-link programmer. If connected and IF in bootloader (if not it won't show up at all), you should see it using lsusb, does it show up?. 

If you have dfu-util 0.8 or higher you should be able to flash it just as any other RIOT Board, although I did change the Makefile of the board for this to work. I pre-seted the dfu ID to 0483:df11 and the default port is /dev/ttyUSB0, both specified in Makefile.include.

**Comment by aabadie** [comment]
Thanks for the clarification about DFU @fjmolinas, I don't have the hardware with me today. I'll try again next week. I'm running dfu-util 0.9 so it should be ok I guess.

FYI, I have a branch rebased and adapted to latest master. If you agree, I can open a PR with this one, and then you can test. Is it ok for you ?

**Comment by fjmolinas** [comment]
@aabadie  Yes, please go ahead.

**Comment by aabadie** [comment]
@fjmolinas thanks, I opened #6850 so you can have a look.

**Comment by kYc0o** [comment]
I'll close this one, reopen if necessary.
